### PR TITLE
Support `artifacts` section

### LINKED ISSUE #519: make_dvcyaml: Create `artifacts` section
Per https://github.com/iterative/dvc/pull/9220

### PR DESCRIPTION
**Pull request by daavoo**
Closes #519

### CODE REVIEW

File: src/dvclive/live.py
Comment (dberenbaum) [inline src/dvclive/live.py:326]: Why not include `type` and `name` as explicit kwargs?

Comment (dberenbaum) [reply src/dvclive/live.py:326]: And maybe even `desc` and `labels`?

**Comment by dberenbaum** [comment]
Thanks @daavoo!

Questions I have:
* Not a blocker, but this reminds me we need to revisit the lightning callback since it's most popular but doesn't auto-log the model artifact.
* Do we know what the artifact names look like for each framework? Should we be inferring them from the path or trying to come up with some better name for them?
* If we log the artifact at the end of each step/epoch, will it raise an error because another artifact exists with the same name?

File: src/dvclive/live.py
Comment (daavoo) [reply src/dvclive/live.py]: I added explicit names. The reasons were that it was less code and `type` is a reserved keyword in Python, but neither are strong reasons

**Comment by dberenbaum** [comment]
> * Do we know what the artifact names look like for each framework? Should we be inferring them from the path or trying to come up with some better name for them?

This would be a breaking change, so can we at least see what the artifact names look like and make a decision on which way to go here?

**Comment by dberenbaum** [comment]
@daavoo We agreed that we would drop the callback support and make this depend on explicit calls to `log_artifact` for now, right? What's the status?

File: src/dvclive/live.py
Comment (aguschin) [inline src/dvclive/live.py:352]: If I written to `dvclive/dvc.yaml` already, I assume `name`s from there won't make it here, right?

Cautious that they can break the workflow: you log a single model, but the first time you run the script you create `model.pkl`, next time `models/model.pkl`, next time `dvclive/models/model.pkl`, until you run out of names.

File: src/dvclive/utils.py
Comment (aguschin) [inline src/dvclive/utils.py:136]: For now we have a regex checking for names https://github.com/iterative/dvc/blob/main/dvc/repo/artifacts.py#L16

we need to make sure dvclive can't produce names that will be invalidated by that regexp. Depending on when you want to merge this, we can revisit this later.

**Review by aguschin — APPROVED** [review]
Looks great overall (and simple - I like that very much after mlem/dvc codebases 😅 ). I see that Dave already posted feedback - have only one few items to add on top of that.

File: src/dvclive/live.py
Comment (daavoo) [reply src/dvclive/live.py:352]: > If I written to dvclive/dvc.yaml already, I assume names from there won't make it here, right?

We don't load the existing `dvclive/dvc.yaml` on instantiation and we overwrite it with the data logged on the current runtime.

This is not only because of `artifacts` but also makes sense for other sections. For example, a user might remove the code that logs parameters so keeping that section in `dvclive/dvc.yaml` would not make sense/break.

> Cautious that they can break the workflow: you log a single model, but the first time you run the script you create model.pkl, next time models/model.pkl, next time dvclive/models/model.pkl, until you run out of names.

As per the above explanation, this should not happen unless the user calls `live.log_artifact` with the same name inside the same script, at which point I don't think we should try to work around it but consider it a misuse of the API.

File: src/dvclive/utils.py
Comment (daavoo) [reply src/dvclive/utils.py:136]: DVCLive now requires DVC Python API so I think I would update the P.R. to import the name check function from DVC and use it here. Does that sound good to you?

**Comment by daavoo** [comment]
> @daavoo We agreed that we would drop the callback support and make this depend on explicit calls to `log_artifact` for now, right? What's the status?

Dropped from the P.R.

File: src/dvclive/dvc.py
Comment (dberenbaum) [inline src/dvclive/dvc.py:117]: Why are we not keying `_models` by `name`? Then I think this becomes a one-line like `dvcyaml["artifacts"] = live._models`.

File: src/dvclive/utils.py
Comment (dberenbaum) [inline src/dvclive/utils.py:132]: Do we need it to be "unmatched"? For example, if I'm doing my own checkpointing, I might repeatedly call `log_artifact` with the same path and I'd expect it to overwrite each time.

File: src/dvclive/dvc.py
Comment (daavoo) [reply src/dvclive/dvc.py:117]: > Why are we not keying _models by name? Then I think this becomes a one-line like dvcyaml["artifacts"] = live._models.

I had it like that and then had to change it, but I don't recall the reason 😅 . 
I will take an extra look

File: src/dvclive/utils.py
Comment (daavoo) [reply src/dvclive/utils.py:132]: I think this was the reason for not using `name` as the key  (answer in the above comment 😅) .

What you describe works as expected in the current P.R. because we use `path` as the key and then check:

```
if path not in self._models:
```

To decide whether a `name` needs to be generated or not

File: src/dvclive/live.py
Comment (daavoo) [reply src/dvclive/live.py:351]: We could, I don't know if we should. 
My motivation was to not include unnecessary/unrelated things in the P.R.

It would not be that hard to extend in the future if/when we need it.

File: src/dvclive/utils.py
Comment (daavoo) [reply src/dvclive/utils.py:132]: To clarify, we could achieve a similar thing using `name`, but the code felt worse to me.
Perhaps I am biased because I still feel that `path` made more sense as primary key in `artifacts` 😅

File: src/dvclive/live.py
Comment (dberenbaum) [reply src/dvclive/live.py:351]: I guess I worry it's harder to document this way. How do you explain what the other artifact metadata args do?

Not a blocker.

File: src/dvclive/utils.py
Comment (dberenbaum) [reply src/dvclive/utils.py:132]: Yeah, it feels to me like the code would be simpler if we use the name key and also don't care about unmatched names. It could infer a name if needed, then add whatever metadata is provided to `_models` as is, and write out `_models` as is to the artifacts section of `dvc.yaml`.

Also not really a blocker, but maybe worth a look since I wonder if the current implementation is biased too much by all the previous discussions and we will struggle to remember why it looks this way when we need to touch it in the future.

Comment (daavoo) [reply src/dvclive/utils.py:132]: > also don't care about unmatched names

Do you mean that we should not worry about:

```py
with Live() as live:
    live.log_artifact("classification/model.pth", type="model")
    live.log_artifact("detection/model.pth", type="model")
```

File: src/dvclive/live.py
Comment (daavoo) [reply src/dvclive/live.py]: > How do you explain what the other artifact metadata args do?

Write to `.dvc` file?

File: src/dvclive/utils.py
Comment (daavoo) [reply src/dvclive/utils.py]: Updated the P.R. to use `name` as primary key

File: src/dvclive/live.py
Comment (daavoo) [reply src/dvclive/live.py]: Side note. 

`artifacts` section has a restriction on `names` formatting that is only applicable because of the model registry + `git tag` usage.

It feels harder to document that we would try to write to `artifacts` except if the `name` we found or the one you provide doesn't fulfill a restriction that doesn't apply to your use case.

File: src/dvclive/utils.py
Comment (daavoo) [reply src/dvclive/utils.py]: > Updated the P.R. to use `name` as primary key

TBH, it feels to me that we are moving the code from one place to another, not really simplifying

File: src/dvclive/live.py
Comment (dberenbaum) [reply src/dvclive/live.py]: > They are written to .dvc file? (I updated the P.R. to do that)

I don't think we should have two different places to add these fields. I'd rather deprecate putting them into `.dvc` files like we are doing with `artifacts.yaml`. This feels like we are only partially committing to the new MR workflow. Is there some hesitation to rely on this new `artifacts` section?

> It feels harder to document that we would try to write to `artifacts` except if the `name` we found or the one you provide doesn't fulfill a restriction that doesn't apply to your use case.

What other use cases do you have in mind?

File: src/dvclive/dvc.py
Comment (dberenbaum) [inline src/dvclive/dvc.py:200]: I think we should overwrite and leave this up to the user to manage, so we could reduce this to `models[name] = attrs`.

File: src/dvclive/dvc.py
Comment (dberenbaum) [inline src/dvclive/dvc.py:207]: Yeah, this looks bad unless we can condense the logic more 😓 . @aguschin WDYT? Can we explain it better? Do we really need it to be this convoluted?

File: src/dvclive/dvc.py
Comment (dberenbaum) [inline src/dvclive/dvc.py:190]: I think we can drop `find_unused_name`. If users are worried about conflicts, they can explicitly set a name.

File: src/dvclive/utils.py
Comment (dberenbaum) [reply src/dvclive/utils.py]: Thanks for trying it this way @daavoo. Left some comments about where I think we can simplify.

File: src/dvclive/dvc.py
Comment (daavoo) [reply src/dvclive/dvc.py:200]: The thing is that `attrs` here is coming from the explicit optional arguments in `log_artifact` that we have for each of the attributes. 

So, here or elsewhere we need a loop like this to discard `None` values, unless we are ok with dumping to `artifacts.yaml` having attributes set to None.

(Not having to handle this was one of the "pros" of `kwargs`)

File: src/dvclive/dvc.py
Comment (daavoo) [reply src/dvclive/dvc.py:190]: I included it because it was explicitly mentioned as a requirement in our last sync with @aguschin and @omesser .
I am happy to go ahead and remove code if there are no strong objections

**Comment by daavoo** [comment]
> Sorry for so many iterations here @daavoo. The blocker left for me is what to do with attrs if there isn't type=model. IMO we should write any attrs to artifacts. It's simple and even if it doesn't do much now, neither does anything else we do with those attrs (at least it could be useful in the future and an artifact registry would work with the existing artifacts written to artifacts).

Pushed a new version

**Review by dberenbaum — APPROVED** [review]

File: tests/test_log_artifact.py
Comment (daavoo) [inline tests/test_log_artifact.py:57]: TODO: Fix the `path` here.
It should be written as relative to the `dvclive/dvc.yaml`

Comment (daavoo) [reply tests/test_log_artifact.py]: I pushed a fix

**Comment by daavoo** [comment]
Don't see much point in merging until we test end-to-end against Studio.

**Comment by aguschin** [comment]
@daavoo, looks like it will be working, but if you want me, I can test this against what I have now. WDYT?

**Comment by daavoo** [comment]
> @daavoo, looks like it will be working, but if you want me, I can test this against what I have now. WDYT?

Sounds good, what just saying that we can wait for merge and release until it works with Studio prod

**Comment by dberenbaum** [comment]
Can we merge and push people to start using it now, letting them know what's coming? We have people today asking to get started with the model registry. I'd like to tell them to use this and their models will appear once Studio is updated. Otherwise, I can only tell them do it the old way but it will break soon, or they can't start using it at all.

**Pull request merged**
